content/en/blog/2021-12-06-fake-it-when-you-make-it.md

```diff
@@ -0,0 +1,262 @@
+<!--more-->
+
+## Testing An Existing Codebase
+
+As developers we'd like to think that we're following test-driven development or at least "development with tests". However, sometimes things need to be built quickly or we find ourselves inheriting an existing codebase. How do we write tests without fundamentally changing the source code?
+
+At Rotational we are interested in building intelligent distributed systems. This adds an additional layer to our testing problem because such projects often involve many external dependencies. A common strategy for dealing with these dependencies is to use mocks.
+
+From a black box perspective a mock imitates the behavior of a service. We use a very generic definition of a service here; in reality it could be an internal or external API, a web server, an external process, or maybe even just a method. At the end of the day we want to be able to test that our code handles the possible scenarios that such as service introduces.
+
+## Go Forth And Mock
+
+In languages like Python, adding unit tests to existing code is fairly straightforward, mostly due to the availability of third party libraries such as [pytest](docs.pytests.org). In addition, the weakly typed nature of Python lends itself to easy mocking. For example, the [unittest.mock](https://docs.python.org/3/library/unittest.mock.html) library can be used to replace method calls with calls that returned mocked responses (e.g., 200 OK responses to HTTP requests).
+
+This is in stark constrast with Go, which has some properties which make it more of a challenge to write effective tests. Go is strongly typed, which makes it impossible to swap out variables and methods for mocked types on the fly. In addition, Go makes an important distinction between exported and unexported variables, an analogue to public and private variables in other languages. While this design is useful for protecting variables from unwanted access, it also makes tests more difficult to implement since they usually exist in separate packages from the source code.
+
+Nevertheless, Go is a popular language for distributed systems and sometimes we need a way to mock things to enable testing. Thankfully, it's possible to create mocks with some effort. Suppose we have an `EmailManager`[^1] struct with an associated method that's responsible for sending emails using the [SendGrid](https://sendgrid.com/) service.
+
+```go
+type EmailConfig struct {
+    SendGridAPIKey string
+}
+
+type EmailManager struct {
+	client sendgrid.Client
+}
+
+func New(conf EmailConfig) (m *EmailManager) {
+    m = &EmailManager{
+        client = sendgrid.NewSendClient(conf.SendGridAPIKey)
+    }
```

Comment:
Return `m` or maybe just some ellipses to suggest that the function continues?

Reply:
Thanks for catching this!